evm_loader/program/src/account/state.rs

```diff
@@ -400,3 +387,200 @@ impl<'a> StateAccount<'a> {
+        let header = super::header::<Header>(&self.account);
+        self.map_obj(header.evm_machine_offset)
+    }
+
+    #[must_use]
+    pub fn read_executor_state(&self) -> ManuallyDrop<Boxx<ExecutorStateData>> {
+        let header = super::header::<Header>(&self.account);
+        self.map_obj(header.executor_state_offset)
+    }
+
+    fn map_obj<T>(&self, offset: usize) -> ManuallyDrop<Boxx<T>> {
+        let data = self.account.data.borrow().as_ptr();
+        unsafe {
+            let ptr = data.add(offset).cast_mut();
+            assert_eq!(ptr.align_offset(align_of::<T>()), 0);
+            let data_ptr = ptr.cast::<T>();
+
+            ManuallyDrop::new(Boxx::from_raw_in(data_ptr, acc_allocator()))
+        }
+    }
+}
+
+impl<'a> StateAccount<'a> {
+    /// Implementation to squeeze bits of information from the state account.
+    /// N.B.
+    /// 1. `StateAccount` contains objects and pointers allocated by the state account allocator, so reading
+    ///     objects inside requires jumping on the offset (between the real account address as allocated by the
+    ///     current allocator) and "intended" address of the first account as provided by the Solana runtime.
+    /// 2. `addr_of!` and `read_unaligned` is heavily used to facilitate the reading of fields by raw pointers.
+    /// 3. There are upcasts from *const u8 to *const T, but since T was allocated by the allocator previously,
+    ///     it has the correct alignment and the upcast is sound.
+    #[allow(clippy::cast_ptr_alignment)]
+    pub fn get_state_account_view(
+        program_id: &Pubkey,
+        account: &AccountInfo<'a>,
+    ) -> Result<StateAccountCoreApiView> {
+        super::validate_tag(program_id, account, TAG_STATE)?;
+
+        let header = super::header::<Header>(account);
+        let memory_space_delta = {
+            account.data.borrow().as_ptr() as isize
+                - isize::try_from(crate::allocator::STATE_ACCOUNT_DATA_ADDRESS)?
+        };
+        // Pointer to the Data is needed to get pointers to the fields in a safe way (using addr_of!).
+        let data_ptr = unsafe {
+            account
+                .data
+                .borrow()
+                .as_ptr()
+                .add(header.data_offset)
+                .cast::<Data>()
+                .cast_mut()
+        };
+
+        unsafe {
+            // Reading full `Transaction`.
+            let transaction_ptr = addr_of!((*data_ptr).transaction);
+            // Memory layout for transaction payload is: tag of enum's variant (usize) followed by the variant value.
+            let tx_payload_ptr = addr_of!((*transaction_ptr).transaction).cast::<usize>();
+            let payload_ptr = tx_payload_ptr.add(1);
+
+            // TODO: ugly manual piece of code that likely can be substituted with the macro.
+            let tx_payload: TransactionPayload = match read_unaligned(tx_payload_ptr) {
```

Comment:
I do need your opinion and attention here @anton-lisanin @s-medvedev.

Let me briefly explain what's happening here. 

**Problem:**
This piece of logic is responsible to reconstruct Transaction from State Account in the context of get_holder Core API call. Because all the transaction objects were allocated in the context of Solana, all the dynamically allocated object have the address around 0x40000000000 (standard solana memory space for all data of all Solana accounts as passed by the solana runtime).
Now, because we read the same State Account from get_holder state, it has different allocator and different memory space (dependent on the system allocator). So, to read the full transaction object, we have to do manual jumps in the memory and reads by exact addresses.

**Solution:**
**Now**:
As of now, all the fields are being performed by `read_unaligned` with the offset that corresponds to the alignment of every field (calldata requires a bit more magic, because it's a vector).

**Potential refactoring:**
I can create a special single purpose proc_derive_macro just for this case and annotate LegacyTx and AccessTx with that. The logic of proc_macro_derive will be to read all the fields of the struct using `read_unaligned` and invoke constructor directly. This will allow to remove the ugly piece at lines 494 - 525. I even have a sketch of it somewhere (so it shouldn't take a long time). WDYT? 

My only little concern - it will be single purpose macro that will be used only in one place and the amount of code will be the same :)

**So the question:**
- do I need to crate such proc_macro_derive?  
- if yes, should I put it in the "**program-macro / evm-loader-macro**" crate or create a new one (rust does not currently allow same-crate proc_macro_derives)?

Reply:
Sure, go ahead with a macro.
You can add it to the existing `program-macro` crate.

Reply:
Done, resolving.